# Reglas de Arquitectura

Este directorio contiene las reglas de arquitectura que se aplican autom√°ticamente a todo el proyecto para mantener consistencia, calidad y mejores pr√°cticas.

## üìã **Reglas Disponibles:**

### **üî§ Nomenclatura y Documentaci√≥n**

- **[english-naming-spanish-docs.mdc](./english-naming-spanish-docs.mdc)**: Nomenclatura en ingl√©s con documentaci√≥n en espa√±ol
  - Define que los nombres de componentes, archivos, variables y c√≥digo deben estar en ingl√©s
  - La documentaci√≥n y comentarios deben estar en espa√±ol
  - Garantiza consistencia internacional del c√≥digo con accesibilidad local

### **‚ö° Calidad de C√≥digo TypeScript**

- **[typescript-best-practices.mdc](./typescript-best-practices.mdc)**: Mejores pr√°cticas de TypeScript
  - Previene errores comunes como par√°metros no utilizados, interfaces incorrectas
  - Define uso correcto de `readonly`, prefijos `_`, y verificaci√≥n de interfaces externas
  - Establece est√°ndares para type safety y mantenibilidad del c√≥digo

### **üß© Integraci√≥n con Tech-Block-Lib**

- **[check-tech-block-lib-first.mdc](./check-tech-block-lib-first.mdc)**: Consultar tech-block-lib antes de crear componentes
  - Mapeo autom√°tico de palabras clave a componentes tech-block-lib
  - Proceso obligatorio de consulta de documentaci√≥n
  - Evita duplicaci√≥n de componentes existentes

### **üèóÔ∏è Arquitectura Angular**

- **[delegation-to-services.mdc](./delegation-to-services.mdc)**: Delegar l√≥gica a servicios
  - Mantiene los componentes limpios delegando l√≥gica de negocio a servicios
  - Promueve la separaci√≥n de responsabilidades

- **[no-direct-dom.mdc](./no-direct-dom.mdc)**: Evitar acceso directo al DOM
  - Usar `Renderer2` o `ElementRef` para mantener compatibilidad con Angular y SSR
  - Evita manipulaci√≥n directa del DOM

- **[use-onpush-strategy.mdc](./use-onpush-strategy.mdc)**: ChangeDetectionStrategy.OnPush
  - Mejora el rendimiento usando la estrategia OnPush para la detecci√≥n de cambios
  - Optimiza el ciclo de detecci√≥n de cambios de Angular

## üéØ **C√≥mo Funcionan las Reglas:**

### **Aplicaci√≥n Autom√°tica:**

Todas las reglas tienen `alwaysApply: true`, lo que significa que se aplican autom√°ticamente cuando trabajas en archivos que coinciden con los patrones definidos en `globs`.

### **Cobertura de Archivos:**

Las reglas cubren diferentes tipos de archivos:

- **TypeScript**: `**/*.ts`
- **HTML Templates**: `**/*.html`
- **Estilos**: `**/*.scss`, `**/*.css`
- **JavaScript**: `**/*.js`
- **Configuraci√≥n**: `**/*.json`
- **Documentaci√≥n**: `**/*.md`

### **Tipos de Reglas:**

- **comment**: Reglas que proporcionan comentarios y gu√≠as durante el desarrollo
- **architecture**: Reglas que definen patrones estructurales del proyecto

## üìö **Beneficios:**

### **‚úÖ Consistencia:**

- Nomenclatura estandarizada en todo el proyecto
- Patrones de c√≥digo unificados
- Estructura de archivos coherente

### **‚úÖ Calidad:**

- Prevenci√≥n de errores comunes
- Mejores pr√°cticas aplicadas autom√°ticamente
- C√≥digo m√°s mantenible y legible

### **‚úÖ Productividad:**

- Reducci√≥n del tiempo de revisi√≥n de c√≥digo
- Gu√≠as autom√°ticas durante el desarrollo
- Integraci√≥n optimizada con tech-block-lib

### **‚úÖ Escalabilidad:**

- C√≥digo compatible con est√°ndares internacionales
- Facilita la incorporaci√≥n de nuevos desarrolladores
- Soporte para equipos distribuidos

## üîß **Agregar Nuevas Reglas:**

Para crear una nueva regla de arquitectura:

1. **Crear archivo `.mdc`** en este directorio
2. **Definir front matter YAML:**
   ```yaml
   ---
   id: nombre-unico-regla
   name: Nombre descriptivo de la regla
   description: Descripci√≥n detallada de qu√© hace la regla
   globs:
     - "**/*.ts"
     - "**/*.html"
   type: comment
   alwaysApply: true
   ---
   ```
3. **Escribir contenido en Markdown** con ejemplos y explicaciones
4. **Actualizar este README** agregando la nueva regla

## üìñ **Estructura de una Regla:**

Cada archivo `.mdc` debe contener:

### **Front Matter (YAML):**

- `id`: Identificador √∫nico
- `name`: Nombre descriptivo
- `description`: Explicaci√≥n breve
- `globs`: Patrones de archivos afectados
- `type`: Tipo de regla (comment, architecture, etc.)
- `alwaysApply`: Si se aplica autom√°ticamente

### **Contenido (Markdown):**

- Explicaci√≥n detallada de la regla
- Ejemplos de c√≥digo correcto e incorrecto
- Casos de uso y excepciones
- Beneficios y justificaci√≥n

## ‚ö†Ô∏è **Reglas Cr√≠ticas:**

Las siguientes reglas son **especialmente importantes** y afectan directamente la arquitectura del proyecto:

1. **english-naming-spanish-docs**: Base para toda la nomenclatura del proyecto
2. **typescript-best-practices**: Previene errores comunes de TypeScript y mejora la calidad del c√≥digo
3. **check-tech-block-lib-first**: Evita duplicaci√≥n de componentes y asegura consistencia con el sistema de dise√±o

**Estas reglas deben seguirse estrictamente para mantener la calidad y consistencia del c√≥digo.**
