# Development Rules / Reglas de Desarrollo

Esta carpeta contiene reglas relacionadas con **errores comunes de desarrollo** y **buenas prÃ¡cticas de TypeScript** que previenen problemas recurrentes durante el desarrollo.

## ğŸ“ Archivos incluidos:

### `typescript-common-errors.mdc`
**Regla preventiva** para evitar errores frecuentes de TypeScript que rompen la compilaciÃ³n y causan problemas de desarrollo.

**Errores que previene:**
- âœ… **Imports incorrectos**: Validators de 'tech-block-lib' en lugar de '@angular/forms'
- âœ… **Imports no utilizados**: Variables importadas pero nunca usadas
- âœ… **Propiedades inexistentes**: dataQaId en interfaces donde no corresponde
- âœ… **Tipos mal escritos**: ILibTbDynamicFormConfig vs ILibTbDynamicFormConfigType
- âœ… **Variables no utilizadas**: Propiedades declaradas pero nunca referenciadas

**CaracterÃ­sticas:**
- âœ… GuÃ­a completa de imports por librerÃ­a (Angular, RxJS, tech-block-lib)
- âœ… Soluciones especÃ­ficas para cada tipo de error
- âœ… Templates de correcciÃ³n rÃ¡pida
- âœ… Checklist de verificaciÃ³n antes de commit
- âœ… ConfiguraciÃ³n recomendada de VSCode/IDE

## ğŸ¯ Objetivo

**Prevenir errores de compilaciÃ³n** y **mejorar la productividad** del desarrollo evitando problemas recurrentes que rompen el build y requieren debugging innecesario.

## ğŸ“‹ CÃ³mo usar

### **Durante desarrollo:**
1. **Antes de crear imports** â†’ Consultar guÃ­a de librerÃ­as correctas
2. **DespuÃ©s de escribir cÃ³digo** â†’ Verificar que todos los imports se usan
3. **Al usar interfaces** â†’ Verificar propiedades vÃ¡lidas
4. **Antes de commit** â†’ Ejecutar `ng build` para detectar errores

### **Para resoluciÃ³n rÃ¡pida:**
1. **Error de import** â†’ Buscar la librerÃ­a correcta en la guÃ­a
2. **Import no usado** â†’ Eliminar o usar inmediatamente
3. **Propiedad inexistente** â†’ Verificar documentaciÃ³n de la interfaz
4. **Tipo mal escrito** â†’ Usar nombres exactos de tech-block-lib

## âš ï¸ Nota importante

Estas reglas fueron creadas basÃ¡ndose en **errores reales** encontrados durante el desarrollo. Son **100% prevenibles** con verificaciÃ³n adecuada y seguimiento de las guÃ­as establecidas.

## ğŸ”§ Comandos Ãºtiles

```bash
# Detectar errores de compilaciÃ³n
ng build --configuration development

# Detectar problemas de linting
ng lint --fix

# Organizar imports automÃ¡ticamente (VSCode)
Shift + Alt + O
```

## ğŸ“š Resumen

| Tipo de Error | PrevenciÃ³n | Comando de VerificaciÃ³n |
|-------|-------|-------|
| **Imports incorrectos** | Consultar guÃ­a de librerÃ­as | `ng build` |
| **Imports no usados** | Organizar imports regularmente | `ng lint` |
| **Propiedades inexistentes** | Verificar interfaces antes de usar | `ng build` |
| **Tipos mal escritos** | Copy-paste nombres exactos | `ng build` |
| **Variables no usadas** | Code review antes de commit | `ng lint` |

**REGLA DE ORO**: Siempre ejecutar `ng build --configuration development` antes de hacer commit para detectar errores temprano. 